lndclient/router_client.go

```diff
@@ -0,0 +1,161 @@
+)
+
+// RouterClient exposes payment functionality.
+type RouterClient interface {
+	SendPayment(ctx context.Context, invoice string,
+		maxFee btcutil.Amount, maxCltv *int32, outgoingChannel *uint64,
+		timeout time.Duration) (chan PaymentStatus, chan error, error)
+
+	TrackPayment(ctx context.Context, hash lntypes.Hash) (
+		chan PaymentStatus, chan error, error)
+}
+
+// PaymentStatus describe the state of a payment.
+type PaymentStatus struct {
+	State    routerrpc.PaymentState
+	Preimage lntypes.Preimage
+	Fee      lnwire.MilliSatoshi
+	Route    []uint64
+}
+
+type routerClient struct {
+	client       routerrpc.RouterClient
+	routerKitMac serializedMacaroon
+}
+
+func newRouterClient(conn *grpc.ClientConn,
+	routerKitMac serializedMacaroon) *routerClient {
+
+	return &routerClient{
+		client:       routerrpc.NewRouterClient(conn),
+		routerKitMac: routerKitMac,
+	}
+}
+
+func (r *routerClient) SendPayment(ctx context.Context, invoice string,
+	maxFee btcutil.Amount, maxCltv *int32, outgoingChannel *uint64,
+	timeout time.Duration) (chan PaymentStatus, chan error, error) {
+
+	rpcCtx, cancel := context.WithTimeout(ctx, rpcTimeout)
+	defer cancel()
+
+	rpcCtx = r.routerKitMac.WithMacaroonAuth(rpcCtx)
+	req := &routerrpc.SendPaymentRequest{
+		FeeLimitSat:    int64(maxFee),
+		PaymentRequest: invoice,
+		TimeoutSeconds: int32(timeout.Seconds()),
+	}
+	if maxCltv != nil {
+		req.CltvLimit = *maxCltv
+	}
+	if outgoingChannel != nil {
+		req.OutgoingChanId = *outgoingChannel
+	}
+
+	stream, err := r.client.SendPayment(rpcCtx, req)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return r.trackPayment(ctx, stream)
+}
+
+func (r *routerClient) TrackPayment(ctx context.Context,
+	hash lntypes.Hash) (chan PaymentStatus, chan error, error) {
+
+	ctx = r.routerKitMac.WithMacaroonAuth(ctx)
+	stream, err := r.client.TrackPayment(
+		ctx, &routerrpc.TrackPaymentRequest{
+			PaymentHash: hash[:],
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return r.trackPayment(ctx, stream)
+}
+
+func (r *routerClient) trackPayment(ctx context.Context,
+	stream routerrpc.Router_TrackPaymentClient) (chan PaymentStatus,
+	chan error, error) {
+
+	statusChan := make(chan PaymentStatus)
+	errorChan := make(chan error, 1)
+	go func() {
+		for {
+			rpcStatus, err := stream.Recv()
+			if err != nil {
+				switch status.Convert(err).Code() {
+				case codes.NotFound:
+					err = channeldb.ErrPaymentNotInitiated
+
+				case codes.AlreadyExists:
+					err = channeldb.ErrAlreadyPaid
+				}
+
+				errorChan <- err
+				return
+			}
+
+			status, err := unmarshallPaymentStatus(rpcStatus)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+
+			select {
+			case statusChan <- *status:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return statusChan, errorChan, nil
+}
+
+func unmarshallPaymentStatus(rpcStatus *routerrpc.PaymentStatus) (
+	*PaymentStatus, error) {
+
+	status := PaymentStatus{
+		State: rpcStatus.State,
```

Comment:
What about the returned errors?

Reply:
Failed payment are just another payment state. Other errors like "payment not found" are already handled above as a stream err.